governance/contract.go

```diff
@@ -0,0 +1,110 @@
+package governance
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/params"
+)
+
+var errContractEngineNotReady = errors.New("ContractEngine is not ready")
+
+type ContractEngine struct {
+	chainConfig   *params.ChainConfig
+	currentParams *params.GovParamSet
+
+	chain blockChain // To access the contract state DB
+}
+
+func NewContractEngine(config *params.ChainConfig) *ContractEngine {
+	e := &ContractEngine{
+		chainConfig:   config,
+		currentParams: params.NewGovParamSet(),
+	}
+
+	return e
+}
+
+func (e *ContractEngine) SetBlockchain(chain blockChain) {
+	e.chain = chain
+}
+
+// Params effective at upcoming block (head+1)
+func (e *ContractEngine) Params() *params.GovParamSet {
+	return e.currentParams
+}
+
+// Params effective at requested block (num)
+func (e *ContractEngine) ParamsAt(num uint64) (*params.GovParamSet, error) {
+	if e.chain == nil {
+		logger.Error("Invoked ParamsAt() before SetBlockchain", "num", num)
+		return nil, errContractEngineNotReady
+	}
+
+	head := e.chain.CurrentHeader().Number.Uint64()
+	if num > head {
+		// Sometimes future blocks are requested.
+		// ex) reward distributor in istanbul.engine.Finalize() requests ParamsAt(head+1)
+		// ex) governance_itemsAt(num) API requests arbitrary num
+		// In those cases we refer to the head block.
+		num = head + 1
+	}
```

Comment:
Since `contractGetAllParamsAt(num)` can accept any future block numbers, we can remove this if-clause.
Just like `governance_itemsAt` can preview future param set, Let's allow `ParamsAt` to preview future param set.

Reply:
done